src/frontend/src/session/cursor_manager.rs

```diff
@@ -0,0 +1,272 @@
+                }
+
+                let timestamp_row: i64 = new_row
+                    .get(0)
+                    .unwrap()
+                    .as_ref()
+                    .map(|bytes| std::str::from_utf8(bytes).unwrap().parse().unwrap())
+                    .unwrap();
+
+                if timestamp_row != self.rw_timestamp {
+                    // 2b. Find next_rw_timestamp, need update cursor with next_rw_timestamp.
+                    return Ok(CursorRowValue::QueryWithNextRwTimestamp(
+                        timestamp_row,
+                        self.subscription_name.clone(),
+                    ));
+                } else {
+                    // 2c. The rw_timestamp of this row is equal to self.rw_timestamp, return row
+                    return Ok(CursorRowValue::Row((
+                        Row::new(build_row_with_logstore(new_row, timestamp_row)?),
+                        self.pg_desc.clone(),
+                    )));
+                }
+            }
+        }
+    }
+}
+
+pub fn build_row_with_snapshot(row: Vec<Option<Bytes>>, rw_timestamp: i64) -> Vec<Option<Bytes>> {
+    let mut new_row = vec![
+        Some(Bytes::from(
+            convert_logstore_i64_to_epoch(rw_timestamp).to_string(),
+        )),
+        Some(Bytes::from(1i16.to_string())),
+    ];
+    new_row.extend(row);
+    new_row
+}
+
+pub fn build_row_with_logstore(
+    mut row: Vec<Option<Bytes>>,
+    rw_timestamp: i64,
+) -> Result<Vec<Option<Bytes>>> {
+    // remove sqr_id, vnode ,_row_id
+    let mut new_row = vec![Some(Bytes::from(
+        convert_logstore_i64_to_epoch(rw_timestamp).to_string(),
+    ))];
+    new_row.extend(row.drain(3..row.len() - 1).collect_vec());
+    Ok(new_row)
+}
+
+pub fn build_desc(mut descs: Vec<PgFieldDescriptor>, is_snapshot: bool) -> Vec<PgFieldDescriptor> {
+    let mut new_descs = vec![
+        PgFieldDescriptor::new(
+            "rw_timestamp".to_owned(),
+            DataType::Varchar.to_oid(),
+            DataType::Varchar.type_len(),
+        ),
+        PgFieldDescriptor::new(
+            "op".to_owned(),
+            DataType::Int16.to_oid(),
+            DataType::Int16.type_len(),
+        ),
+    ];
+    if is_snapshot {
+        new_descs.extend(descs)
+    } else {
+        new_descs.extend(descs.drain(4..descs.len() - 1));
+    }
+    new_descs
+}
+
+pub enum CursorRowValue {
+    Row((Row, Vec<PgFieldDescriptor>)),
+    QueryWithNextRwTimestamp(i64, ObjectName),
+    QueryWithStartRwTimestamp(i64, ObjectName),
+}
+#[derive(Default)]
+pub struct CursorManager {
+    cursor_map: HashMap<String, (Cursor, Instant)>,
+    // Save subsciption's retentain_secs
+    cursor_retention_secs_maps: HashMap<ObjectName, Duration>,
+}
+
+impl CursorManager {
+    pub fn add_cursor_retention_secs(
+        &mut self,
+        subscription_name: ObjectName,
+        retention_secs: Duration,
+    ) {
+        self.cursor_retention_secs_maps
+            .insert(subscription_name, retention_secs);
```

Comment:
Discussed offline. The assumptions here for `cursor_retention_secs_maps` are:
- `CursorManager` is per-session. That means states in `CursorManager` will be cleaned when session is destroyed.
- All provided `retention_secs` in `add_cursor_retention_secs` for the same `subscription_name` are the same because cursors in the same session share the same session timeout and subscription retention seconds cannot be change.


Correct me if I am wrong, based on these assumptions, I think we don't need to maintain `cursor_retention_secs_maps` because:
1. When session timeout kicks-in, the cursor will be dropped anyway. That means `retention_secs` in `add_cursor_retention_secs` can be `subscription` retention only without considering session timeout.
2. `subscription` retention can be retrived via the session catalog. That means `cursor_retention_secs_maps` stores redundant information.
3. There is no need to check whether the cursor has been inactive for more than `subscription` retention on fetch cursor calls. The check is only neccessary when user specify `rw_timestamp` in declare cursor and `rw_timestamp` falls out of retention.